pythonforandroid/bootstraps/common/build/src/main/java/org/kivy/android/DashCamPythonService.java

```diff
@@ -0,0 +1,212 @@
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            notification = new Notification(
+                context.getApplicationInfo().icon, serviceTitle, System.currentTimeMillis());
+            try {
+                // prevent using NotificationCompat, this saves 100kb on apk
+                Method func = notification.getClass().getMethod(
+                    "setLatestEventInfo", Context.class, CharSequence.class,
+                    CharSequence.class, PendingIntent.class);
+                func.invoke(notification, context, serviceTitle, serviceDescription, pIntent);
+            } catch (NoSuchMethodException | IllegalAccessException |
+                     IllegalArgumentException | InvocationTargetException e) {
+            }
+        } else {
+            // for android 8+ we need to create our own channel
+            // https://stackoverflow.com/questions/47531742/startforeground-fail-after-upgrade-to-android-8-1
+            String NOTIFICATION_CHANNEL_ID = "org.kivy.p4a";    //TODO: make this configurable
+            String channelName = "Background Service";                //TODO: make this configurable
+            NotificationChannel chan = new NotificationChannel(NOTIFICATION_CHANNEL_ID, channelName, 
+                NotificationManager.IMPORTANCE_NONE);
+            
+            chan.setLightColor(Color.BLUE);
+            chan.setLockscreenVisibility(Notification.VISIBILITY_PRIVATE);
+            NotificationManager manager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+            manager.createNotificationChannel(chan);
+
+            Notification.Builder builder = new Notification.Builder(context, NOTIFICATION_CHANNEL_ID);
+            builder.setContentTitle(serviceTitle);
+            builder.setContentText(serviceDescription);
+            builder.setContentIntent(pIntent);
+            builder.setSmallIcon(context.getApplicationInfo().icon);
+            notification = builder.build();
+        }
+        startForeground(getServiceId(), notification);
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        pythonThread = null;
+        if (autoRestartService && startIntent != null) {
+            Log.v("python service", "service restart requested");
+            startService(startIntent);
+        }
+        Process.killProcess(Process.myPid());
+    }
+
+    /**
+     * Stops the task gracefully when killed.
+     * Calling stopSelf() will trigger a onDestroy() call from the system.
+     */
+    @Override
+    public void onTaskRemoved(Intent rootIntent) {
+        super.onTaskRemoved(rootIntent);
+        stopSelf();
+    }
+
+    @Override
+    public void run(){
+        String app_root =  getFilesDir().getAbsolutePath() + "/app";
+        File app_root_file = new File(app_root);
+        PythonUtil.loadLibraries(app_root_file,
+            new File(getApplicationInfo().nativeLibraryDir));
+        this.mService = this;
```

Comment:
Do you know why this block cannot simply be in this service `onCreate`? so that python executes within the thread of this service? Python will create other threads as it needs to.

Reply:
This is a copy cat of PythonService but contain fixes only to handle service crashes.